EventFilter/Phase2TrackerRawToDigi/plugins/DTHDAQToFEDRawDataConverter.cc

```diff
@@ -0,0 +1,145 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include <fstream>
+#include <vector>
+#include <iostream>
+#include <iomanip>
+#include <cstdint>
+
+// Constants for bit field widths, markers, and sizes
+constexpr unsigned int orbitHeaderSize = 32;
+constexpr unsigned int fragmentTrailerSize = 16;
+constexpr unsigned int fragmentPayloadWordSize = 16;  // Each fragment payload word is 16 bytes
+constexpr uint8_t orbitHeaderMarkerH = 0x48;
+constexpr uint8_t orbitHeaderMarkerO = 0x4F;
+constexpr uint8_t fragmentTrailerMarkerF = 0x48;
+constexpr uint8_t fragmentTrailerMarkerH = 0x46;
+
+class DTHDAQToFEDRawDataConverter : public edm::stream::EDProducer<> {
+public:
+  explicit DTHDAQToFEDRawDataConverter(const edm::ParameterSet&);
+  ~DTHDAQToFEDRawDataConverter() override = default;
+
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+private:
+  std::string inputFile_;
+  unsigned int fedId_;
+
+  std::vector<char> readRawFile(const std::string& inputFile);
+  void parseAndDumpEventData(const std::vector<char>& buffer);
+  void reverseParseFragments(const std::vector<char>& buffer, size_t startIdx, uint32_t packetWordCount, uint16_t eventCount);
+  void printHex(const std::vector<char>& buffer, size_t length);
+};
+
+DTHDAQToFEDRawDataConverter::DTHDAQToFEDRawDataConverter(const edm::ParameterSet& config)
+  : inputFile_(config.getParameter<std::string>("inputFile")),
+    fedId_(config.getParameter<unsigned int>("fedId")) {
+  produces<FEDRawDataCollection>();
+}
+
+std::vector<char> DTHDAQToFEDRawDataConverter::readRawFile(const std::string& inputFile) {
+  std::ifstream rawFile(inputFile, std::ios::binary | std::ios::ate);
+  if (!rawFile.is_open()) {
+    throw cms::Exception("FileOpenError") << "Could not open input file: " << inputFile;
+  }
+
+  std::streamsize fileSize = rawFile.tellg();
+  rawFile.seekg(0, std::ios::beg);
+
+  std::vector<char> buffer(fileSize);
+  if (!rawFile.read(buffer.data(), fileSize)) {
+    throw cms::Exception("FileReadError") << "Could not read input file: " << inputFile;
+  }
+
+  rawFile.close();
+  return buffer;
+}
+
+void DTHDAQToFEDRawDataConverter::printHex(const std::vector<char>& buffer, size_t length) {
+  std::cout << "Raw bitstream (first " << length << " bytes): ";
+  for (size_t i = 0; i < std::min(buffer.size(), length); ++i) {
+    std::cout << std::hex << std::setw(2) << std::setfill('0')
+              << (static_cast<unsigned int>(static_cast<unsigned char>(buffer[i]))) << " ";
+  }
+  std::cout << std::dec << std::endl;
+}
+
+void DTHDAQToFEDRawDataConverter::parseAndDumpEventData(const std::vector<char>& buffer) {
+  size_t orbitSize = buffer.size() / 4;  // Divide buffer for 4 orbits
```

Comment:
Why 4 orbits?

Reply:
I changed it to constant (orbitCount) so the user can adjust the number of orbits. I used 4 because this is (for now) the number of orbits we expect to have per file from the TIF people (as far as I know)